src/main/cpp/include/genomicsdb/genomicsdb_iterators.h

```diff
@@ -0,0 +1,310 @@
+ * the Software without restriction, including without limitation the rights to 
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
+ * the Software, and to permit persons to whom the Software is furnished to do so, 
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all 
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef GENOMICSDB_ITERATORS_H
+#define GENOMICSDB_ITERATORS_H
+
+#include "headers.h"
+#include "variant_array_schema.h"
+#include "genomicsdb_columnar_field.h"
+#include "vid_mapper.h"
+#include "c_api.h"
+#include "timer.h"
+
+//Exceptions thrown
+class GenomicsDBIteratorException : public std::exception {
+  public:
+    GenomicsDBIteratorException(const std::string m="") : msg_("GenomicsDBIteratorException exception : "+m) { ; }
+    ~GenomicsDBIteratorException() { ; }
+    // ACCESSORS
+    /** Returns the exception message. */
+    const char* what() const noexcept { return msg_.c_str(); }
+  private:
+    std::string msg_;
+};
+
+//Whenever a live cell must be pointed to in a buffer, use this structure
+//Useful for the left sweep, tracking cells while producing VCF records etc
+class GenomicsDBLiveCellMarker
+{
+  public:
+    GenomicsDBLiveCellMarker(const size_t num_markers, const unsigned num_fields)
+    {
+      m_valid.resize(num_markers);
+      m_initialized.resize(num_markers);
+      m_row.resize(num_markers);
+      m_begin.resize(num_markers);
+      m_end.resize(num_markers);
+      for(auto i=0ull;i<num_markers;++i)
+      {
+        m_buffer_ptr_vec.emplace_back(num_fields, static_cast<GenomicsDBBuffer*>(0));
+        m_indexes.emplace_back(num_fields);
+      }
+      reset();
+    }
+    inline void reset()
+    {
+      //Set columns to -1
+      m_begin.assign(m_begin.size(), -1ll);
+      m_initialized.assign(m_initialized.size(), false);
+      m_valid.assign(m_valid.size(), false);
+    }
+    inline void set_row_idx(const size_t idx, const int64_t row_idx)
+    {
+      assert(idx < m_row.size());
+      m_row[idx] = row_idx;
+    }
+    inline int64_t get_row_idx(const size_t idx) const
+    {
+      assert(idx < m_row.size());
+      return m_row[idx];
+    }
+    inline void set_column_interval(const size_t idx, const int64_t b, const int64_t e)
+    {
+      assert(idx < m_begin.size() && idx < m_end.size());
+      m_begin[idx] = b;
```

Comment:
if these two always have same index, maybe use a std::pair<int64_t, int64_t> to store them